### ISSUE TITLE
Import error while trying to parse irrelevant command line arguments

### ISSUE THREAD

**Issue by mzedayda** [datapane/datapane#182]
### System Information

- OS: Mac OS
- Python version: 3.7.7
- Python environment: pip
- Using jupyter: false
- Datapane version: 0.12.3

### Bug / Issue
Hi there,  we recently started to use datapane on our team to generate reports and we really like it.

We encountered an issue in our CI.
Whenever we import datapane in one of our tests, we get the following error: 
`click.exceptions.BadParameter: 'no:warnings', should be name=value`

The reason is we run our tests using:
`python -m pytest tests/ -p no:warnings`

But datapane in its main __init__.py module tries to parse any command line parameter starting with '-p', and in our case fails because this flag is for pytest's use.

I'm not sure what the correct behavior should be, one option is to disable this override mechanism when datapane is used as a lib (and not an explicit datapane command), another option is to use more specific flag prefix instead of just '-p', maybe '--datapane-p' or something similar.

Hope this issue is not a duplicate, I didn't find anything similar.

Full error stack:
```
/usr/local/lib/python3.7/site-packages/datapane/client/utils.py:104: in split_param
    k, v = x.split("=", maxsplit=1)
E   ValueError: not enough values to unpack (expected 2, got 1)
During handling of the above exception, another exception occurred:
tests/reporting/test_report_creator.py:8: in <module>
    from reporting.serve_report.report_creator import _get_report_pages_block
reporting/serve_report/report_creator.py:4: in <module>
    import datapane as dp
/usr/local/lib/python3.7/site-packages/datapane/__init__.py:87: in <module>
    config = parse_command_line()
/usr/local/lib/python3.7/site-packages/datapane/client/utils.py:132: in parse_command_line
    return process_cmd_param_vals(dp_args.parameter)
/usr/local/lib/python3.7/site-packages/datapane/client/utils.py:111: in process_cmd_param_vals
    return {k: convert_param_val(v) for (k, v) in split_param(params)}
/usr/local/lib/python3.7/site-packages/datapane/client/utils.py:111: in <dictcomp>
    return {k: convert_param_val(v) for (k, v) in split_param(params)}
/usr/local/lib/python3.7/site-packages/datapane/client/utils.py:106: in split_param
    raise click.BadParameter(err_msg.format(x))
E   click.exceptions.BadParameter: 'no:warnings', should be name=value
```

**Comment by mands** [comment]
Hi @mzedayda ,

Thanks for raising this - we'll take a look and see if we can resolve this, unsure why it's trying to parse the command-line when being used as a library, but should be possible to reproduce and a simple fix.

**Comment by igormoochnick** [comment]
It fails for me as well. It's impossible to use DataPane library in a regular application that has it's own command line parameters. The parameter parsing logic in __init__.py can't be circumvented because it's activated on `import datapane as dp`. My application has it's own command line parameters and it obviously has nothing to do with the ones that datapane is expecting

**Comment by shahinrostami** [comment]
To keep everyone in the loop – we'll be looking into this next week 👍

**Comment by shahinrostami** [comment]
I've reproduced this with a simple test:

```python
def test_datapane_report():
    import datapane as dp
    import os

    report = dp.Report("# My report")
    report.save(path="my_report.html")
    assert os.path.exists("my_report.html") == False
```


```
shahin@datapane ~/Documents/reproduce_182 % poetry run python3 -m pytest tests/ -p no:warnings
================================================================================== test session starts ==================================================================================
platform darwin -- Python 3.10.5, pytest-7.1.3, pluggy-0.13.1
rootdir: /Users/shahin/Documents/reproduce_182
collected 2 items                                                                                                                                                                       

tests/test_reproduce_182.py .F                                                                                                                                                    [100%]

======================================================================================= FAILURES ========================================================================================
_________________________________________________________________________________ test_datapane_report __________________________________________________________________________________

xs = ['no:warnings']

    def split_param(xs: Tuple[str, ...]) -> Iterator[t.Tuple[str, str]]:
        err_msg = "'{}', should be name=value"
        for x in xs:
            try:
>               k, v = x.split("=", maxsplit=1)
E               ValueError: not enough values to unpack (expected 2, got 1)

.venv/lib/python3.10/site-packages/datapane/client/utils.py:125: ValueError

During handling of the above exception, another exception occurred:

    def test_datapane_report():
>       import datapane as dp

tests/test_reproduce_182.py:9: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
.venv/lib/python3.10/site-packages/datapane/__init__.py:132: in <module>
    config = parse_command_line()
.venv/lib/python3.10/site-packages/datapane/client/utils.py:153: in parse_command_line
    return process_cmd_param_vals(dp_args.parameter)
.venv/lib/python3.10/site-packages/datapane/client/utils.py:132: in process_cmd_param_vals
    return {k: convert_param_val(v) for (k, v) in split_param(params)}
.venv/lib/python3.10/site-packages/datapane/client/utils.py:132: in <dictcomp>
    return {k: convert_param_val(v) for (k, v) in split_param(params)}
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

xs = ['no:warnings']

    def split_param(xs: Tuple[str, ...]) -> Iterator[t.Tuple[str, str]]:
        err_msg = "'{}', should be name=value"
        for x in xs:
            try:
                k, v = x.split("=", maxsplit=1)
            except ValueError:
>               raise click.BadParameter(err_msg.format(x))
E               click.exceptions.BadParameter: 'no:warnings', should be name=value

.venv/lib/python3.10/site-packages/datapane/client/utils.py:127: BadParameter
================================================================================ short test summary info ================================================================================
FAILED tests/test_reproduce_182.py::test_datapane_report - click.exceptions.BadParameter: 'no:warnings', should be name=value
============================================================================== 1 failed, 1 passed in 8.18s ==============================================================================
```

**Comment by mands** [comment]
Just to provide a bit more background information here on the intended use-case (pinging @shahinrostami and @spookylukey also)

The main code is in [`parse_command_line`](https://github.com/datapane/datapane/blob/master/projects/python-client/src/datapane/client/utils.py#L135) and is only used for parsing parameters that can then be loaded into the dp.Config object as a way to paramterise your script as it runs (i.e. it's not actually used by datapane itself, but a feature to help report writers pass data into their reports when running in batch mode, i.e. in a CI system or ML pipeline).

I think there seems to be a bug in the handling here, but also we should use a more descriptive argument name, like `--dp-parameter` to avoid name-clashes.

**Comment by shahinrostami** [comment]
This was automatically closed after merging a PR that solves the issue.

I''m re-opening it and marking it as `release pending` until the fix makes it into a release.

**Comment by spookylukey** [comment]
@mands @shahinrostami  The normal mechanism for doing that in scripts would be environment variables, IMO it's really surprising that a library would ever parse the command line like we are doing here:

https://github.com/datapane/datapane/blob/4cead8674b30dfb296054940385c62c60e7bf98d/projects/python-client/src/datapane/__init__.py#L127-L133

It seems like the only way to fix bug this is to break compatibility with existing scripts which use datapane as a library and expect this behaviour (as was done in #277 - scripts expecting the old behaviour won't work). Since we have a compatibility break in that case, we may as well switch to a better way of doing it i.e. specify a set of `DATAPANE_*` environment variables that can be used for this purpose. 

In addition, #277 breaks existing CLI usage without any good reason — for the explicit command line case, we've got no reason to change anything, as we in are control of what we expect, so we shouldn't be breaking people's scripts (i.e. scripts that start `datapane ... --parameter=...`)

Also, if we try to fix it with different command line parameter names, we're still going to have issues with people trying to do it that if they use other tools that are expecting to consume the whole command line. For example, if someone is using the standard library module [argparse](https://docs.python.org/3/library/argparse.html) to build up the command line options for their own entry point script, it's going to crash out with `unrecognized arguments: --dp-parameter` if they try to pass that as well. Almost all well written scripts that parse arguments in a standard way (such as pytest etc.) will have this behaviour - they reject arguments that they don't understand. This means the CLI arguments for configuring library behaviour is going to be unsatisfactory in the majority of cases.

So in short, I think we should:
1. revert #277 so that our CLI usage remains the same
2. remove https://github.com/datapane/datapane/blob/4cead8674b30dfb296054940385c62c60e7bf98d/projects/python-client/src/datapane/__init__.py#L132
3. add and document configuration method based on environment variables which is used only for library case

I'm happy to work on this if you'd like

**Comment by spookylukey** [comment]
Looking further, and thinking about the use case:

To support script writers who want a script with the same interface as `datapane` CLI, we don't want environment variables at all. We simply need to expose/document the `parse_command_line() ` function so that people can call it themselves if they want it.

So, it seems the right fix is to simply remove the call to `parse_command_line` when this is being called as a function, and document somewhere that it can be called manually if you want it.

**Comment by mands** [comment]
Thanks for the comment here.

I agree that there may be issues when users are using datapane as a library with their own command-line parser, and that is a concern we should avoid. (The "library" terminology we've used internally is perhaps not the most helpful, ideally we'd run datapane as a server, like gunicorn/express or even how `python -m pytest` does, but we need to support the use-case of people iterating within Jupyter which makes that a harder goal to achieve).

I'm personally not keen on the env var solution for a couple of reasons,
- It isn't the convention on Windows, where it's really awkward to setup custom env vars to use when running a script, and is far from a common practice
- Many users will run the scripts directly from within Jupyter, using a `!` shell command, where env vars may not work as intended
- I worry that some more introductory Data Scientists would find this complex, where they may be familiar with the concept of command-line arguments but not environment variables

Regarding the feature itself, I don't believe it's used very heavily (if at all, afaik it's not actually documented), so we should be ok to change it - we'll be doing a minor release soon as part of #275 that will be breaking BC so we can release both in tandem.

> To support script writers who want a script with the same interface as datapane CLI, we don't want environment variables at all. We simply need to expose/document the parse_command_line() function so that people can call it themselves if they want it.
> So, it seems the right fix is to simply remove the call to parse_command_line when datapane is being used as a library, and document somewhere that it can be called manually if you want it.

This does seem like a good outcome, i.e. making it opt-in rather than automatic, and we can export something like `dp.parse_command_line()` for users to call at the top of their script if desired. (The exact way the feature works may be redundant with the introduction of Functions - but some similar will still be required imo).

Would you be able to make a PR?

**Comment by spookylukey** [comment]
@mands Yep I'll make a PR.

**Comment by shahinrostami** [comment]
Re-opening until fix makes it into release 🚀

**Comment by shahinrostami** [comment]
This should be fixed in our latest release 🎉

I've confirmed the fix with the following test:

```python
def test_datapane_report():
    import datapane as dp
    import os

    app = dp.App("# My app")
    app.save(path="my_app.html")
    assert os.path.exists("my_app.html") == True
```

<img width="836" alt="Screenshot 2022-09-22 at 19 27 31" src="https://user-images.githubusercontent.com/15690380/191823622-cfbcf8a0-37d6-4a1c-8444-0f9412f10de0.png">

Thank you for raising this – please let us know if there's anything else.

**Comment by mzedayda** [comment]
Thank you for your support 🙏
